Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainMacroTextureClipmapComputePass.cpp

```diff
@@ -0,0 +1,221 @@
+        {
+            m_clipmapData.m_clipmapCenters[clipmapIndex].SetX(m_clipmapData.m_clipmapCenters[clipmapIndex].GetZ());
+            m_clipmapData.m_clipmapCenters[clipmapIndex].SetY(m_clipmapData.m_clipmapCenters[clipmapIndex].GetW());
+
+            const AZ::Vector2 scaledTranslation = normalizedViewTranslation * clipmapScale;
+            // If normalized translation on a certain level of clipmap goes out of the current clipmap representation,
+            // a full update will be triggered and the center will be reset to the center.
+            if (AZStd::abs(scaledTranslation.GetX()) >= 1.0f || AZStd::abs(scaledTranslation.GetY()) >= 1.0f)
+            {
+                m_clipmapData.m_fullUpdateFlag |= (1 << clipmapIndex);
+                m_clipmapData.SetCurrentClipmapCenter(AZ::Vector2(0.5f, 0.5f), clipmapIndex);
+            }
+            else
+            {
+                AZ::Vector2 clipmapCenter = AZ::Vector2(
+                    m_clipmapData.m_clipmapCenters[clipmapIndex].GetZ(),
+                    m_clipmapData.m_clipmapCenters[clipmapIndex].GetW());
+
+                clipmapCenter = clipmapCenter + scaledTranslation;
+
+                float centerX = clipmapCenter.GetX();
+                float centerY = clipmapCenter.GetY();
+
+                AZ_Assert((centerX > -1.0f) && (centerX < 2.0f) && (centerY > -1.0f) && (centerY < 2.0f),
+                    "The new translated clipmap center should be within this range. Otherwise it should fall into the other if branch and reset the center to (0.5, 0.5).");
+
+                // Equivalent to modulation.
+                // It wraps around to use toroidal addressing.
+                centerX -= centerX > 1.0f ? 1.0f : 0.0f;
+                centerX += centerX < 0.0f ? 1.0f : 0.0f;
+
+                centerY -= centerY > 1.0f ? 1.0f : 0.0f;
+                centerY += centerY < 0.0f ? 1.0f : 0.0f;
+
+                clipmapCenter = AZ::Vector2(centerX, centerY);
+
+                m_clipmapData.SetCurrentClipmapCenter(clipmapCenter, clipmapIndex);
+            }
+
+            clipmapScale *= 2.0f;
+        }
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::ClipmapData::SetPreviousClipmapCenter(const AZ::Vector2& clipmapCenter, uint32_t clipmapLevel)
+    {
+        m_clipmapCenters[clipmapLevel].SetX(clipmapCenter.GetX());
+        m_clipmapCenters[clipmapLevel].SetY(clipmapCenter.GetY());
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::ClipmapData::SetCurrentClipmapCenter(const AZ::Vector2& clipmapCenter, uint32_t clipmapLevel)
+    {
+        m_clipmapCenters[clipmapLevel].SetZ(clipmapCenter.GetX());
+        m_clipmapCenters[clipmapLevel].SetW(clipmapCenter.GetY());
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::ClipmapData::SetMaxRenderSize(const AZ::Vector3& maxRenderSize)
+    {
+        memcpy(&m_maxRenderSize, &maxRenderSize, sizeof(m_maxRenderSize));
```

Comment:
I don't think this is safe and Vector3 has a function for this - I think it's something like `StoreToFloatArray`